### PR TITLE
Fix for change of billing plan

### PR DESCRIPTION
**Pull request by 4patelr**
Fixes Issue 47 by canceling existing subscription and creating a new subscription in Stripe when the billing plan for a Tenant is updated.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by 4patelr** [comment]
#47 - Fixes issue 47

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java [hunk 1]
```diff
@@ -271,58 +272,94 @@ private void provisionTenantInStripe(String tenantId, String planId) throws Stri
...
+                LOGGER.info("provisionTenantInStripe: Customer created with id: " + customer.getId());
+            } catch (StripeException e) {
+                LOGGER.error("provisionTenantInStripe: Error creating Customer for Tenant: " + tenantId);
+                LOGGER.error(Utils.getFullStackTrace(e));
+                throw e;
+            }
+        } else {
+            LOGGER.info("provisionTenantInStripe: Customer: {} create skipped, already exists with id: {} ", customerName, customer.getId());
+            //If subscription plan has changed then cancel the old subscriptions and create a new.
+            //Ideally, the subscription information for the tenant would be stored in a DDB table so we do not have to hit Stripe to find if it has changed
```
Comment (PoeppingT) [inline metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:316]: If the customer can change subscription information directly in Stripe, we should *not* store any subscription info in DDB as that information may become inconsistent. We can build optimizations around Stripe (or other billing providers) if we find that it becomes a bottleneck.

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java [hunk 2]
```diff
@@ -271,58 +272,94 @@ private void provisionTenantInStripe(String tenantId, String planId) throws Stri
...
-        LOGGER.info("provisionTenantInStripe: Setting up Tenant: " + customerId + " as a customer and creating subscription");
+        final String customerName = "tenant-" + tenantId.split("-")[0];  //get first segment
+        LOGGER.info("provisionTenantInStripe: Setting up Tenant: {} as a customer and creating subscription", customerName);
 
-        try {
-            //Only attempt to setup tenant if customer does not already exist for this tenant
-            try {
-                Customer customer = Customer.retrieve(customerId);
-                LOGGER.info("Customer : " + customerId + " already exists, setup of tenant: " + tenantId + " skipped");
+        Customer customer = getStripeCustomer(customerName);
```
Comment (PoeppingT) [inline metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:283]: What happens if Stripe is down or times out when responding? Do we fail the entire tenant onboarding stack? Should we retry, either immediately or lazily?

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java [hunk 3]
```diff
@@ -271,58 +272,94 @@ private void provisionTenantInStripe(String tenantId, String planId) throws Stri
...
+                            LOGGER.info("Subscription Plan has not changed for the tenant, nothing to do");
+                            return;
+                        }
+                    }
+                    //NOTE: Do queued metered events need to sent to Stripe before canceling the subscription?
+
+                    //attempt to cancel the subscription
+                    LOGGER.info("provisionTenantInStripe: No match in Stripe subscription items for plan {}", planId);
+                    LOGGER.info("provisionTenantInStripe: Attempt to Cancel Subscription: {} ", subscription.getId());
+                    Subscription canceledSubscription = subscription.cancel();
```
Comment (PoeppingT) [inline metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:337]: Why not [update the subscription](https://stripe.com/docs/api/subscriptions/update) instead of cancelling and recreating?

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java [hunk 4]
```diff
@@ -366,25 +388,47 @@ private void cancelSubscriptionInStripe(String tenantId) throws StripeException
...
 
     /*
-    Creates a customer and a subscription in Stripe
+    Creates a subscription in Stripe
     The Subscription will have one item for the monthly plan and additional items for metered products
     For the metered products, a message is sent to Event Bus to add that record to the Dynanamo table mapping
      the Tenant internal product code to the Stripe Subscription id which is unique for each tenant and product.
      */
-    private void createStripeSubscription(Customer customer, SubscriptionPlan plan, String fullTenantId) throws StripeException{
+    private void createStripeSubscription(Customer customer, SubscriptionPlan plan, String fullTenantId, boolean existingCustomer) throws StripeException{
```
Comment (PoeppingT) [inline metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:431]: We're now passing in `existingCustomer` but it doesn't appear you're using it.

File: resources/custom-resources/disable-instance-protection/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/DisableInstanceProtection.java [hunk 5]
```diff
@@ -55,8 +55,8 @@ public Object handleRequest(Map<String, Object> event, Context context) {
         LOGGER.info("Processing for Autoscaling group {}", autoScalingGroup);
         try {
             Runnable r = () -> {
-                if ("Delete".equalsIgnoreCase(requestType)) {
-                    LOGGER.info("DELETE");
+                if ("Delete".equalsIgnoreCase(requestType) || "Update".equalsIgnoreCase(requestType)) {
+                    LOGGER.info("Request Type: " + requestType);
```
Comment (PoeppingT) [inline resources/custom-resources/disable-instance-protection/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/DisableInstanceProtection.java:59]: These changes don't appear to be related to fixing billing. Can you explain why this and the change to saas-boost-svc-onboarding.yaml are in the diff?

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java (hunk 1 above)
Comment (PoeppingT) [reply metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:316]: In this case by customer I meant the SaaS vendor/builder/ISV using Boost, as I would assume they have the highest level of control eventually over the Stripe subscription of behalf of the tenant; after all, the Stripe subscription is so the ISV can bill the tenant for usage.

I'm fine with sticking with this current implementation, I'm just suggesting storing information in DDB may have some unforeseen side effects.

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java (hunk 3 above)
Comment (PoeppingT) [reply metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:337]: Both great questions. Sounds like when we pick up this PR we'll need to do some experimentation with Stripe's API to see how it acts in different situations.

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java (hunk 2 above)
Comment (brtrvn) [reply metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:283]: Agree that we should have some sort of async retry mechanism in place for 3rd party integrations, but we are currently safe from blowing up the entire tenant onboarding/tenant update flow since we invoke these billing service api's async currently.

File: resources/custom-resources/disable-instance-protection/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/DisableInstanceProtection.java (hunk 5 above)
Comment (4patelr) [reply resources/custom-resources/disable-instance-protection/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/DisableInstanceProtection.java:59]: The current flow is when a tenant is updated in UI, it triggers a CloudFormation Stack Update. This change is necessary to send the message to the BillingService with the Billing Plan in case the Billing Plan has changed. The Billing Service will determine if the plan has changed.

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java (hunk 3 above)
Comment (4patelr) [reply metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:337]: Upon further research, Stripe does allow updating the subscription with a new plan and it will do the Proration. 
Take a look at https://stripe.com/docs/billing/subscriptions/change and https://stripe.com/docs/billing/subscriptions/prorations. May be we need to look at this approach instead of the cancel and create new subscription.

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java (hunk 2 above)
Comment (4patelr) [reply metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java:283]: I thought if the Lambda errors out the message from EventBridge will be retried.

File: metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java (hunk 4 above)
Comment (4patelr) [reply metering-billing/lambdas/src/main/java/com/amazon/aws/partners/saasfactory/metering/onboarding/BillingIntegration.java]: Updated to remove unused variable
